Gems/Atom/RPI/Code/Source/RPI.Reflect/Material/MaterialPropertyCollection.cpp

```diff
@@ -0,0 +1,383 @@
+            }
+            else if (value.Is<Vector2>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Vector2>());
+            }
+            else if (value.Is<Vector3>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Vector3>());
+            }
+            else if (value.Is<Vector4>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Vector4>());
+            }
+            else if (value.Is<Color>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Color>());
+            }
+            else if (value.Is<Data::Instance<Image>>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Data::Instance<Image>>());
+            }
+            else if (value.Is<Data::Asset<ImageAsset>>())
+            {
+                return SetPropertyValue(propertyIndex, value.GetValue<Data::Asset<ImageAsset>>());
+            }
+            else
+            {
+                AZ_Assert(false, "Unhandled material property value type");
+                return false;
+            }
+        }
+
+        template<typename Type>
+        const Type& MaterialPropertyCollection::GetPropertyValue(MaterialPropertyIndex index) const
+        {
+            static const Type defaultValue{};
+
+            const MaterialPropertyDescriptor* propertyDescriptor = nullptr;
+            if (Validation::IsEnabled())
+            {
+                if (!index.IsValid())
+                {
+                    AZ_Assert(false, "GetPropertyValue: Invalid MaterialPropertyIndex");
+                    return defaultValue;
+                }
+
+                propertyDescriptor = m_layout->GetPropertyDescriptor(index);
+
+                if (!ValidatePropertyAccess<Type>(propertyDescriptor))
+                {
+                    return defaultValue;
+                }
+            }
+
+            const MaterialPropertyValue& value = m_propertyValues[index.GetIndex()];
+            if (value.Is<Type>())
+            {
+                return value.GetValue<Type>();
+            }
+            else
+            {
+                if (Validation::IsEnabled())
+                {
+                    AZ_Assert(false, "Material property '%s': Stored property value has the wrong data type. Expected %s but is %s.",
+                    propertyDescriptor->GetName().GetCStr(),
+                        azrtti_typeid<Type>().template ToString<AZStd::string>().data(), // 'template' because clang says "error: use 'template' keyword to treat 'ToString' as a dependent template name"
+                        value.GetTypeId().ToString<AZStd::string>().data());
+                }
+                return defaultValue;
+            }
+        }
+
+        // Using explicit instantiation to restrict GetPropertyValue to the set of types that we support
```

Comment:
Minor nit, I feel like I would have just used ADL (argument deduction lookup) for this instead as a way to simplify the code and reading it (and likely have fewer footguns also)

Reply:
I'm not exactly sure what you mean.

But FWIW, this code is just moved from another file, not introduced by this PR. IIRC I original wrote the code this way to allow the function to be defined in the cpp file instead of h file, and also to prevent confusion if a programmer tries to call GetPropertyValue (or SetPropertyValue is similar) with an unsupported type (it would eventually report errors at runtime but why not detect the problem at compile time).